experimental/dds/tree2/src/core/tree/visitDelta.ts

```diff
@@ -3,319 +3,677 @@
+					if (mark.newContent.detachId === undefined) {
+						const nodeId = offsetDetachId(mark.newContent.restoreId, i);
+						const { root: restoredRoot } = config.treeIndex.getEntry(nodeId);
+						const newContent = { source: restoredRoot, nodeId, fields: mark.fields };
+						replace.newContent = newContent;
+					}
+					if (mark.oldContent !== undefined) {
+						const { root } = config.treeIndex.getOrCreateEntry(
+							offsetDetachId(mark.oldContent.detachId, i),
+						);
+						replace.oldContent = { destination: root, fields: mark.oldContent.fields };
+					}
+					return replace;
+				});
+			}
+			break;
+		}
+		default:
+			unreachableCase(type);
+	}
+	return [];
+}
+
+/**
+ * Populates the `config` members with operations that need to be performed on detached roots as part of the detach pass.
+ */
+function catalogDetachPassRootChanges(mark: Exclude<Delta.Mark, number>, config: PassConfig): void {
+	let nodeId: Delta.DetachedNodeId | undefined;
+	let fields: Delta.FieldMarks | undefined;
+	switch (mark.type) {
+		case Delta.MarkType.Insert: {
+			if (mark.content.length > 0) {
+				const root = ensureCreation(mark, config);
+				if (mark.fields !== undefined) {
+					config.rootChanges.set(root, mark.fields);
+				}
+			}
+			break;
+		}
+		case Delta.MarkType.Modify:
+			nodeId = mark.detachedNodeId;
+			fields = mark.fields;
+			break;
+		case Delta.MarkType.MoveOut:
+			nodeId = mark.detachedNodeId;
+			fields = mark.fields;
+			if (nodeId !== undefined) {
+				const minor = extractFromOpaque(mark.moveId);
+				const destinationNodeId = { major: "move", minor };
+				const { root: rootSource } = config.treeIndex.getEntry(nodeId);
+				const { root: rootDestination } =
+					config.treeIndex.getOrCreateEntry(destinationNodeId);
+				addRootReplaces(mark.count, config, rootDestination, rootSource, nodeId);
+			}
+			break;
+		case Delta.MarkType.Restore: {
+			nodeId = mark.newContent.restoreId;
+			fields = mark.fields;
+			break;
+		}
+		default:
+			break;
+	}
+	if (nodeId !== undefined && fields !== undefined) {
+		const { root } = config.treeIndex.getOrCreateEntry(nodeId);
+		config.rootChanges.set(root, fields);
+	}
+}
+
+/**
+ * Populates the `config` members with operations that need to be performed on detached roots as part of the attach pass.
+ */
+function catalogAttachPassRootChanges(mark: Exclude<Delta.Mark, number>, config: PassConfig): void {
+	let nodeId: Delta.DetachedNodeId | undefined;
+	let fields: Delta.FieldMarks | undefined;
+	switch (mark.type) {
+		case Delta.MarkType.Insert: {
+			if (mark.content.length > 0) {
+				const rootSource = ensureCreation(mark, config);
+				nodeId = mark.detachId;
+				fields = mark.fields;
+				if (mark.detachId !== undefined) {
+					const count = mark.content.length;
+					const { root: rootDestination } = config.treeIndex.getOrCreateEntry(
+						mark.detachId,
+						count,
+					);
+					addRootReplaces(count, config, rootDestination, rootSource);
+				}
```

Comment:
Why do we do this here instead of just putting the new nodes in this root field in the detach pass?

Reply:
I think it's because inserts could've been made as part of modifications so we have to do it after the detach and detach pass root changes steps.